src/type_checker/checker.rs

```diff
@@ -141,6 +147,16 @@ impl<B: Backend> TypeChecker<B> {
                 })?;
                 let fn_sig = fn_info.sig().clone();
 
+                // check if the function is a hint
+                if fn_info.is_hint && !unsafe_attr {
+                    return Err(self.error(ErrorKind::ExpectedUnsafeAttribute, expr.span));
+                }
+
+                // unsafe attribute should only be used on hints
+                if !fn_info.is_hint && *unsafe_attr {
+                    return Err(self.error(ErrorKind::UnexpectedUnsafeAttribute, expr.span));
+                }
```

Comment:
maybe more concise:

```rust
match (fn_info.is_hint, unsafe_attr) {
  (true, true) | (false, false) => (),
  _ => return Err(self.error(ErrorKind::ExpectedUnsafeAttribute, expr.span));
  }
```

Reply:
Actually there are two error types to handle: https://github.com/zksecurity/noname/blob/18825a3a4a53572bd038c679e823ffc5d2e17ff9/src/negative_tests.rs#L476-L500